### ISSUE TITLE
 Expose API along the lines of ipfs.cat

### ISSUE THREAD

**Issue by Gozala** [ipfs-shipyard/nft.storage#589]
At the moment niftysave asks private ipfs cluster to fetch a tokenURI when it recognizes it to be a content on IPFS (which happens if URL is ipfs:// style  URL or if it looks like gateway URL).

However as per https://github.com/ipfs-shipyard/nft.storage/issues/326 and relevant discussion at https://filecoinproject.slack.com/archives/C02BZPRS9HP/p1634058155211200 chances are that `cat` will fail and nothing will continue trying to fetch that content.

Which is why I would like to propose exposing and endpoint on web3.storage that would attempt to provide the content (internally it could check own s3 ask cluster etc...). Better yet it could also incorporate stuff we wanted from gateway, namely formatting `dag-cbor` nodes as JSON.

I think this would be a much better place to handle relentless attempts to get content than doing it on niftysave. It would also cut direct cluster dependency from niftysave as it would be able to both pin and fetch content in needs through our service directly.

**Comment by Gozala** [comment]
As a side note niftysave sometimes has a gateway URL it encountered, at might be worth providing it e.g. as `sourceURL=https://nft.yo/ipfs/${cid}` which could be tried as a fallback.

**Comment by Gozala** [comment]
I would really like if interface to that was something like

https://*.storage/ipfs://${cid}/foo/bar

**Comment by Gozala** [comment]
Here is the actionable plan to unblock #635 

Expose new API endpoint `https://api.nft.storage/ipfs://` (I know it looks strange but is valid and immediately intuitive) that

1. Validates auth token to ensure it matches the one issued for niftysave.
2. Extracts `ipfs://` URL and responds with fetch to an IPFS gateway corresponding corresponding to this request.


Please note: In this iteration it is not going to try to handle non dag-pb nodes like dag-cbor, that could be a future improvement.

**Comment by Gozala** [comment]
> this endpoint would also support ipfs path (ipfs:///image.png) ?

Do you meant to have a CID in there ? Otherwise I don’t think that’s a valid ipfs URL.

Paths indeed may be supplied

**Comment by Gozala** [comment]
> Do we support ipns/dnslink here ?

No, I consider those to be ipns:// urls and definitely out of scope here

**Comment by Gozala** [comment]
> Do we have a canonical way to go from IPFS URI to Gateway URL and vice versa ? Like a lib that handles all the possible cases ?

I wrote this module that we probably can factor out into nore reusable piece 

https://github.com/ipfs-shipyard/nft.storage/blob/main/packages/niftysave/src/ipfs-url.js

**Comment by hugomrdias** [comment]
> > this endpoint would also support ipfs path (ipfs:///image.png) ?
> 
> Do you meant to have a CID in there ? Otherwise I don’t think that’s a valid ipfs URL.
> 
> Paths indeed may be supplied

Yes I did GitHub squashed it for some reason

**Comment by JeffLowe** [comment]
closing in favor of #1365
